clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java

```diff
@@ -0,0 +1,256 @@
+        final Map<String, ListOffsetsTopic> topicsByName = new HashMap<>();
+        boolean supportsMaxTimestamp = false;
+        for (TopicPartition topicPartition : keys) {
+            OffsetSpec offsetSpec = offsetSpecsByPartition.get(topicPartition);
+            long offsetQuery = getOffsetFromSpec(offsetSpec);
+            if (offsetQuery == ListOffsetsRequest.MAX_TIMESTAMP) {
+                if (skipMaxTimestampRequests) {
+                    continue;
+                }
+                supportsMaxTimestamp = true;
+            }
+            ListOffsetsTopic topic = topicsByName.computeIfAbsent(
+                topicPartition.topic(), topicName -> new ListOffsetsTopic().setName(topicName));
+            topic.partitions().add(
+                new ListOffsetsPartition()
+                    .setPartitionIndex(topicPartition.partition())
+                    .setTimestamp(offsetQuery));
+        }
+
+        return ListOffsetsRequest.Builder
+            .forConsumer(true, options.isolationLevel(), supportsMaxTimestamp)
+            .setTargetTimes(new ArrayList<>(topicsByName.values()));
+    }
+
+    @Override
+    public ApiResult<TopicPartition, ListOffsetsResultInfo> handleResponse(
+        Node broker,
+        Set<TopicPartition> keys,
+        AbstractResponse abstractResponse
+    ) {
+        ListOffsetsResponse response = (ListOffsetsResponse) abstractResponse;
+        Map<TopicPartition, ListOffsetsResultInfo> completed = new HashMap<>();
+        Map<TopicPartition, Throwable> failed = new HashMap<>();
+        List<TopicPartition> unmapped = new ArrayList<>();
+        Set<TopicPartition> retriable = new HashSet<>();
+
+        for (ListOffsetsTopicResponse topic : response.topics()) {
+            for (ListOffsetsPartitionResponse partition : topic.partitions()) {
+                TopicPartition topicPartition = new TopicPartition(topic.name(), partition.partitionIndex());
+                Errors error = Errors.forCode(partition.errorCode());
+                OffsetSpec offsetRequestSpec = offsetSpecsByPartition.get(topicPartition);
+                if (offsetRequestSpec == null) {
+                    log.warn("ListOffsets response includes unknown topic partition {}", topicPartition);
+                } else if (error == Errors.NONE) {
+                    Optional<Integer> leaderEpoch = (partition.leaderEpoch() == ListOffsetsResponse.UNKNOWN_EPOCH)
+                        ? Optional.empty()
+                        : Optional.of(partition.leaderEpoch());
+                    completed.put(
+                        topicPartition,
+                        new ListOffsetsResultInfo(partition.offset(), partition.timestamp(), leaderEpoch));
+                } else {
+                    handlePartitionError(topicPartition, error, failed, unmapped, retriable);
+                }
+            }
+        }
+
+        maybeMarkOtherFailures(
```

Comment:
What does this method do? we have already handled all partitions.

Reply:
It handles `MAX_TIMESTAMP` partitions (which are not eagerly failed in `handleUnsupportedVersionException` but just omitted from the next request) and sanity-checks all partitions (the equivalent of [this check](https://github.com/apache/kafka/blob/9b409de1e2719f72e3c17663ce131e40dc459361/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L3892-L3900)).